did_core/did_parser_nom/src/did_url/mod.rs

```diff
@@ -0,0 +1,164 @@
+    }
+
+    pub fn namespace(&self) -> Option<&str> {
+        self.namespace
+            .clone()
+            .map(|range| self.did_url[range].as_ref())
+    }
+
+    pub fn id(&self) -> Option<&str> {
+        self.id.clone().map(|range| self.did_url[range].as_ref())
+    }
+
+    pub fn path(&self) -> Option<&str> {
+        self.path.as_ref().map(|path| &self.did_url[path.clone()])
+    }
+
+    pub fn queries(&self) -> HashMap<String, String> {
+        self.queries
+            .iter()
+            .map(|(k, v)| {
+                (
+                    self.did_url[k.clone()].to_string(),
+                    self.did_url[v.clone()].to_string(),
+                )
+            })
+            .collect()
+    }
+
+    pub fn fragment(&self) -> Option<&str> {
+        self.fragment
+            .as_ref()
+            .map(|fragment| &self.did_url[fragment.clone()])
+    }
+
+    // TODO: Ideally we would have a builder instead of purpose-specific constructors
+    pub fn from_fragment(fragment: String) -> Result<Self, ParseError> {
+        if all_consuming(fragment_parser)(&fragment).is_err() {
+            return Err(ParseError::InvalidInput("Invalid fragment"));
+        }
+        let len = fragment.len();
+        Ok(Self {
+            did_url: format!("#{}", fragment),
+            did: None,
+            method: None,
+            namespace: None,
+            id: None,
+            path: None,
+            fragment: Some(1..len + 1),
+            queries: HashMap::new(),
+        })
+    }
+
+    pub(crate) fn from_did_parts(
+        did_url: String,
+        did: DidRange,
+        method: Option<DidRange>,
+        id: DidRange,
+    ) -> Self {
+        Self {
+            did_url,
+            did: Some(did),
+            method,
+            namespace: None,
+            id: Some(id),
+            path: None,
+            fragment: None,
+            queries: HashMap::new(),
+        }
+    }
+}
+
+impl TryFrom<String> for DidUrl {
+    type Error = ParseError;
+
+    fn try_from(did_url: String) -> Result<Self, Self::Error> {
+        Self::parse(did_url)
+    }
+}
+
+impl FromStr for DidUrl {
+    type Err = ParseError;
+
+    fn from_str(did: &str) -> Result<Self, Self::Err> {
+        Self::parse(did.to_string())
+    }
+}
+
+impl Display for DidUrl {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.did_url)
+    }
+}
+
+impl Serialize for DidUrl {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&self.did_url)
+    }
+}
+
+impl<'de> Deserialize<'de> for DidUrl {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let did_url = String::deserialize(deserializer)?;
+        DidUrl::parse(did_url).map_err(serde::de::Error::custom)
+    }
+}
+
+impl TryFrom<&DidUrl> for Did {
```

Comment:
If we are implementing a trait for `Did`, shouldn't that rather be placed under the `did` module?

Reply:
I don't think implementing a trait for a struct necessarily means it should be placed in the same module. Doing what you suggest would mean:

* `did` module would know about `DidUrl`, whereas it currently doesn't (which makes sense, `DidUrl` conceptually "derives" from / elaborates on `Did`),
* `DidUrl` would need to expose multiple private fields it currently doesn't.